x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/RestrictedTrustManagerTests.java

```diff
@@ -92,6 +97,32 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
 
         numberOfClusters = scaledRandomIntBetween(2, 8);
         numberOfNodes = scaledRandomIntBetween(2, 8);
+        fields = randomNonEmptySubsetOf(Set.of(SAN_OTHER_COMMON, SAN_DNS));
+    }
+
+    public void testTrustsOnlyNameDns() throws Exception {
+        final Path cert = getDataPath("/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode_updated.crt");
+        baseTrustManager = CertParsingUtils.getTrustManagerFromPEM(List.of(cert));
+        X509Certificate[] certs = CertParsingUtils.readX509Certificates(Collections.singletonList(cert));
+        assertTrue(certs[0].getSubjectAlternativeNames().stream().filter(pair -> (Integer) pair.get(0) == 0).findAny().isEmpty());
+        certificates.put("onlyDns", certs);
+        List<String> validDnsNames = randomNonEmptySubsetOf(
```

Comment:
Missing dnsName wildcard test coverage? Examples:
1. `dnsName="*.localdomain"` with `trust_restrictions=[ "*.localdomain" ]`
2. `dnsName="*.example.com"` with `trust_restrictions=[ "*.com" ]`
3. `dnsName="*.example.com"` with `trust_restrictions=[ "*.example.*" ]`

Reply:
Missing dnsName ACE encoding test for I18N mentioned in [RFC 5280 Section 7.2](https://datatracker.ietf.org/doc/html/rfc5280#section-7.2).
```
   IA5String is limited to the set of ASCII characters.  To accommodate
   internationalized domain names in the current structure, conforming
   implementations MUST convert internationalized domain names to the
   ASCII Compatible Encoding (ACE) format as specified in Section 4 of
   RFC 3490 before storage in the dNSName field.
```
If design intent is to defer I18N dnsName support, maybe add a test to verify ACE encoded values are rejected or ignored.

Reply:
Is design intent for trust_restrictions to do greedy matching or lazy matching?

SAN dnsName matching is lazy (i.e. one subdomain) per HTTP/TLS [RFC 2818 Section 3.1](https://www.ietf.org/rfc/rfc2818.html#section-3.1).
```
   Matching is performed using the matching rules specified by
   RFC2459.  If more than one identity of a given type is present in
   the certificate (e.g., more than one dNSName name, a match in any one
   of the set is considered acceptable.) Names may contain the wildcard
   character * which is considered to match any single domain name
   component or component fragment. E.g., *.a.com matches foo.a.com but
   not bar.foo.a.com. f*.com matches foo.com but not bar.com.
```

It looks like SAN dnsName logic reuses same code as SAN otherName. I can't tell if SAN otherName design intent was to do lazy or greedy matching based on [Elastic Cloud doc examples](https://www.elastic.co/guide/en/cloud/current/ec-trust-management.html#ec-trust-self-managed). Would it be a good idea to add some more tests to make design intent more clear?

Reply:
The match aligns with our other usage of how match strings. We support wildcard expressions and lucene regex's. Both are string matches, not aware of domain names.  The docs need to be updated (they need it before and after this PR) and will be handled in a different PR (likely in a different repository).

Reply:
The wild test coverage is provided by the existing test methods (specifically [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/RestrictedTrustManagerTests.java#L139)). The certificates were updated to include the same value for both commonName and dnsName so that we can [randomize](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/RestrictedTrustManagerTests.java#L100) which one we use for the match.